File: app/src/main/java/com/app/sid/funwithflags/view/countrylist/CountryLocalDataSource.java
package com.app.sid.funwithflags.view.countrylist;

import com.app.sid.funwithflags.datasets.remote.Country;
import com.app.sid.funwithflags.model.realm.RealmCountry;
import com.app.sid.funwithflags.realm.RealmManager;

import java.util.ArrayList;
import java.util.List;

import io.realm.Realm;
import io.realm.RealmResults;
import rx.Observable;

public class CountryLocalDataSource implements CountryListMvp.LocalDataSource {

    private RealmManager mRealManager;

    public CountryLocalDataSource(RealmManager realmManager) {
        mRealManager = realmManager;
    }

    @Override
    public boolean isLocalDataPresent() {
        return !mRealManager.getRealm().isEmpty();
    }

    @Override
    public Observable<List<Country>> getCountryList() {
        List<Country> mCountries = new ArrayList<>();
        Realm realm = mRealManager.getRealm();
        try {
            RealmResults<RealmCountry> realmCountry = realm.where(RealmCountry.class).findAll();
            if (realmCountry != null) {
                for (int i = 0; i < realmCountry.size(); i++) {
                    mCountries.add(realmCountry.get(i).asApiModel());
                }
            }
            return Observable.just(mCountries);
        } finally {
            mRealManager.closeRealm(realm);
        }
    }

    @Override
    public void saveCountry(Country country) {
        Realm realm = mRealManager.getRealm();
        try {
            realm.executeTransaction(transaction -> {
                RealmCountry realmCountry = transaction.where(RealmCountry.class)
                        .contains("name", country.getName())
                        .findFirst();
                if (realmCountry == null) {
                    RealmCountry realmCountryToStore = new RealmCountry(country);
                    transaction.copyToRealmOrUpdate(realmCountryToStore);
                }
            });
        } finally {
            mRealManager.closeRealm(realm);
        }
    }
}

File: app/src/main/java/com/app/sid/funwithflags/view/countrylist/CountryLocalDataSource.kt
package com.app.sid.funwithflags.view.countrylist

import com.app.sid.funwithflags.datasets.remote.Country
import com.app.sid.funwithflags.model.realm.RealmCountry
import com.app.sid.funwithflags.realm.RealmManager
import io.realm.Realm
import rx.Observable
import java.util.*

class CountryLocalDataSource(private val mRealManager: RealmManager) : CountryListMvp.LocalDataSource {

    override fun isLocalDataPresent(): Boolean {
        return !mRealManager.getRealm().isEmpty
    }

    override fun getCountryList(): Observable<List<Country>> {
        val mCountries = mutableListOf<Country>()
        val realm = mRealManager.getRealm()
        return try {
            realm.where(RealmCountry::class.java).findAll()?.forEach {
                mCountries.add(it.asApiModel())
            }
            Observable.just(mCountries)
        } finally {
            mRealManager.closeRealm(realm)
        }
    }

    override fun saveCountry(country: Country?) {
        val realm = mRealManager.getRealm()
        try {
            realm.executeTransaction { transaction: Realm ->
                val realmCountry = transaction.where(RealmCountry::class.java)
                        .contains("name", country!!.name)
                        .findFirst()
                if (realmCountry == null) {
                    val realmCountryToStore = RealmCountry(country)
                    transaction.copyToRealmOrUpdate(realmCountryToStore)
                }
            }
        } finally {
            mRealManager.closeRealm(realm)
        }
    }

}